### PR TITLE
Implementing user definable behaviour for repeat, shuffle, and crossfade

### PR DESCRIPTION
**Pull request by ashenshugarRET**

### CODE REVIEW

**Comment by ashenshugarRET** [comment]
I've split the previous commit, this only covers the changes to make the repeat, shuffle and crossfade reset behaviour definable in usersettings.json. If these changes pass the testing, I'll submit a pull request for the changes to TuneIn command coding and the addition of a "Favorites" command.

**Comment by ashenshugarRET** [comment]
Hello,

Is there any chance you could try and merge this pull request?

Thanks

**Comment by ryanolf** [comment]
Hi, thanks for your patience. I’ve been pretty busy— I’ll try to look
tonight (pacific time).

**Comment by ashenshugarRET** [comment]
Thanks, I understand.

File: lib/process_sonos_command.js [hunk 1]
```diff
@@ -29,9 +29,12 @@ export default async function process_sonos_command(received_text) {
   } else if (received_text_lower.startsWith('spotify:')) {
     service_type = 'spotify';
     sonos_instruction = 'spotify/now/' + received_text;
-  } else if (received_text_lower.startsWith('tunein:')) {
+  } else if (received_text_lower.startsWith('tunein/')) {
```
Comment (ryanolf) [inline lib/process_sonos_command.js]:
```suggestion
  } else if (received_text_lower.startsWith('tunein:')) {
```

For consistency with other URIs (e.g. Spotify, Apple, let's accept using a colon as delimiter. If you want to ALSO accept a slash, that's fine too. But for consistency NFC card will have a string: `tunein:{station ID}`.

File: lib/process_sonos_command.js [hunk 2]
```diff
@@ -29,9 +29,12 @@ export default async function process_sonos_command(received_text) {
...
     service_type = 'tunein';
-    sonos_instruction = 'tunein/now/' + received_text;
+    sonos_instruction = received_text;
```
Comment (ryanolf) [inline lib/process_sonos_command.js]:
```suggestion
    sonos_instruction = 'tunein/play/' + received_text;
```

I guess I had this wrong before. Checking http-api docs, the correct URI string is ```tunein/play/{station ID}```.

File: lib/process_sonos_command.js [hunk 3]
```diff
@@ -29,9 +29,12 @@ export default async function process_sonos_command(received_text) {
...
+  } else if (received_text_lower.startsWith('favorite/')) {
```
Comment (ryanolf) [inline lib/process_sonos_command.js]:
```suggestion
  } else if (received_text_lower.startsWith('favorite:')) {
```

Again for consistency let's program cards with colon delimiters. OK to also accept "slashes" if you want to add that.

File: lib/process_sonos_command.js [hunk 4]
```diff
@@ -29,9 +29,12 @@ export default async function process_sonos_command(received_text) {
...
+    service_type = 'favorite';
+    sonos_instruction = received_text;
```
Comment (ryanolf) [inline lib/process_sonos_command.js]:
```suggestion
    sonos_instruction = 'favorite/' + received_text;
```

File: lib/__tests__/process_sonos_command.test.js [hunk 5]
```diff
@@ -43,8 +43,13 @@ describe('process_sonos_command', () => {
     },
     {
       service: 'TuneIn',
-      input: `tunein:${identifier}`,
-      expected_url: get_sonos_url(`tunein/now/tunein:${identifier}`, 'tunein'),
+      input: `${identifier}`,
```
Comment (ryanolf) [inline lib/__tests__/process_sonos_command.test.js]:
```suggestion
      input: `tunein:${identifier}`,
```

File: lib/__tests__/process_sonos_command.test.js [hunk 6]
```diff
@@ -43,8 +43,13 @@ describe('process_sonos_command', () => {
...
+      expected_url: get_sonos_url(`${identifier}`, 'tunein'),
```
Comment (ryanolf) [inline lib/__tests__/process_sonos_command.test.js]:
```suggestion
      expected_url: get_sonos_url(`tunein/play/${identifier}`, 'tunein'),
```

File: lib/__tests__/process_sonos_command.test.js [hunk 7]
```diff
@@ -43,8 +43,13 @@ describe('process_sonos_command', () => {
...
+    },
+    {
+      service: 'Favorite',
+      input: `${identifier}`,
```
Comment (ryanolf) [inline lib/__tests__/process_sonos_command.test.js]:
```suggestion
      input: `favorite:${identifier}`,
```

File: lib/__tests__/process_sonos_command.test.js [hunk 8]
```diff
@@ -43,8 +43,13 @@ describe('process_sonos_command', () => {
...
+      expected_url: get_sonos_url(`${identifier}`, 'favorite'),
```
Comment (ryanolf) [inline lib/__tests__/process_sonos_command.test.js]:
```suggestion
      expected_url: get_sonos_url(`favorite/${identifier}`, 'favorite'),
```

**Comment by ryanolf** [comment]
Hi @ashenshugarRET, just suggested a few tweaks in the code. The main thing is that for consistency with the URIs written to the cards for Apple, etc, we should use ":" as a delimiter. I'm happy for you to add "/" as an acceptable option as well if you'd like. I like being permissive.

The other thing is that the tests were not passing on my machine. The CI is still pending I think. Please run the test suite before you update the PR. Thanks!

**Comment by ashenshugarRET** [comment]
Apologises, I'm not super experienced with Github etc, to be honest, I don't know what the CI is or in fact how to run it. or the test suite.

**Comment by ryanolf** [comment]
You can run the test suite on your local repo with `npm test`. I just pulled the updates and tests are still failing. I think I can fix if you give me a minute.

**Comment by ashenshugarRET** [comment]
Thanks, yours is the first repository I've used that has test etc, so a bit of a steep learning curve.
Thanks for your help resolving the problems.

File: lib/process_sonos_command.js [hunk 9]
```diff
@@ -31,7 +31,10 @@ export default async function process_sonos_command(received_text) {
     sonos_instruction = 'spotify/now/' + received_text;
   } else if (received_text_lower.startsWith('tunein:')) {
     service_type = 'tunein';
-    sonos_instruction = 'tunein/now/' + received_text;
+    sonos_instruction = 'tunein/play/' + received_text;
+  } else if (received_text_lower.startsWith('favorite:')) {
+    service_type = 'favorite';
+    sonos_instruction = 'favorite/' + received_text;
```
Comment (ryanolf) [inline lib/process_sonos_command.js]:
```suggestion
    sonos_instruction = 'favorite/' + received_text.slice(9);
```

File: lib/process_sonos_command.js [hunk 10]
```diff
@@ -31,7 +31,10 @@ export default async function process_sonos_command(received_text) {
     sonos_instruction = 'spotify/now/' + received_text;
   } else if (received_text_lower.startsWith('tunein:')) {
     service_type = 'tunein';
-    sonos_instruction = 'tunein/now/' + received_text;
+    sonos_instruction = 'tunein/play/' + received_text;
```
Comment (ryanolf) [inline lib/process_sonos_command.js]:
```suggestion
    sonos_instruction = 'tunein/play/' + received_text.slice(7);
```

**Review by ryanolf — APPROVED** [review]

**Comment by ryanolf** [comment]
Ok, we got it. Thanks for contributing.

**Comment by ashenshugarRET** [comment]
Thanks, the node-sonos-http-api has just added the BBC Sounds service to stream BBC radio stations which I was hoping to add to this repository to as I think it will be very popular.
Would it be OK if I generate another pull request to implement this too?

**Comment by ryanolf** [comment]
Yeah, sure! Please add tests to the test suite for it and run the test
suite before you file the pull requests. You run the tests on your local
machine with `npm test` in the root folder of the project. You can read
more about running tests with npm in their docs.

Ryan

**Pull request merged**